crates/wit-parser/src/ast.rs

```diff
@@ -14,29 +15,129 @@ pub mod toposort;
 
 pub use lex::validate_id;
 
-pub struct Ast<'a> {
+enum Ast<'a> {
+    ExplicitPackages(Vec<ExplicitPackage<'a>>),
+    PartialImplicitPackage(PartialImplicitPackage<'a>),
+}
+
+pub struct ExplicitPackage<'a> {
+    package_id: PackageName<'a>,
+    decl_list: DeclList<'a>,
+}
+
+pub struct PartialImplicitPackage<'a> {
     package_id: Option<PackageName<'a>>,
+    decl_list: DeclList<'a>,
+}
+
+/// Stores all of the declarations in a package's scope. In AST terms, this
+/// means everything except the `package` declaration that demarcates a package
+/// scope. In the traditional implicit format, these are all of the declarations
+/// non-`package` declarations in the file:
+///
+/// ```wit
+/// package foo:name;
+///
+/// /* START DECL LIST */
+/// // Some comment...
+/// interface i {}
+/// world w {}
+/// /* END DECL LIST */
+/// ```
+///
+/// In the explicit package style, a [`DeclList`] is everything inside of each
+/// `package` element's brackets:
+///
+/// ```wit
+/// package foo:name {
+///   /* START FIRST DECL LIST */
+///   // Some comment...
+///   interface i {}
+///   world w {}
+///   /* END FIRST DECL LIST */
+/// }
+///
+/// package bar:name {
+///   /* START SECOND DECL LIST */
+///   // Some comment...
+///   interface i {}
+///   world w {}
+///   /* END SECOND DECL LIST */
+/// }
+/// ```
+pub struct DeclList<'a> {
     items: Vec<AstItem<'a>>,
 }
 
+fn parse_explicit_package_items<'a>(tokens: &mut Tokenizer<'a>) -> Result<DeclList<'a>> {
```

Comment:
Could this, and `parse_implicit_package_items` below, be scoped under a `impl DeclList<'_> { ... }` block?

Reply:
Done.